### PR TITLE
feat: conversation filter component

### PR DESCRIPTION
**Pull request by scmmishra**
This PR adds the new conversation filter component, we need to merge the `feat/new-dropdown` branch before merging this

CleanShot 2024-11-18 at 09 24 56@2x

### CODE REVIEW

**Comment by pranavrajs** [comment]
@scmmishra If I click outside, the dropdown doesn't close properly. Please see the screenshot below. 

Screenshot 2024-11-18 at 5 52 29 PM

Please also note the color contrast between the input box and the modal. I'd test this after the issue is resolved.

**Comment by scmmishra** [comment]
> @scmmishra If I click outside, the dropdown doesn't close properly. Please see the screenshot below.

Seems like this came in when I moved to using the Dialog component, fixing it

**Comment by scmmishra** [comment]
> @scmmishra If I click outside, the dropdown doesn't close properly. Please see the screenshot below.

@pranavrajs fixed this

https://github.com/user-attachments/assets/e7506aa3-126b-493a-ad1b-47357063479c

File: app/javascript/dashboard/components-next/dialog/Dialog.vue [hunk 1]
```diff
@@ -44,21 +44,37 @@ defineProps({
...
+  const classesMap = {
+    '3xl': 'max-w-3xl',
+    '2xl': 'max-w-2xl',
+    xl: 'max-w-xl',
+    lg: 'max-w-lg',
+    md: 'max-w-md',
+    sm: 'max-w-sm',
+  };
+
+  return classesMap[props.width];
```
Comment (scmmishra) [inline app/javascript/dashboard/components-next/dialog/Dialog.vue:75]:
```suggestion
  return classesMap[props.width] ?? 'max-w-md';
```

File: app/javascript/dashboard/i18n/locale/en/components.json [hunk 2]
```diff
@@ -5,8 +5,10 @@
   },
   "COMBOBOX": {
     "PLACEHOLDER": "Select an option...",
+    "EMPTY_SEARCH_RESULTS": "No items found for the search term \"{searchTerm}\"",
```
Comment (pranavrajs) [inline app/javascript/dashboard/i18n/locale/en/components.json:8]:
```suggestion
    "EMPTY_SEARCH_RESULTS": "No items found for the search term `{searchTerm}`",
```

Comment (pranavrajs) [reply app/javascript/dashboard/i18n/locale/en/components.json:8]: The `\` character would create a mess in the translations.

File: app/javascript/dashboard/components-next/icon/ChannelIcon.vue [hunk 3]
```diff
@@ -0,0 +1,17 @@
+<script setup>
```
Comment (pranavrajs) [inline app/javascript/dashboard/components-next/icon/ChannelIcon.vue:1]: Could you create a separate PR for this? It appears unrelated to the filters.

File: app/javascript/dashboard/components-next/button/Button.vue [hunk 4]
```diff
@@ -11,36 +11,60 @@ const props = defineProps({
...
+  trailingIcon: { type: Boolean, default: false },
+  isLoading: { type: Boolean, default: false },
 });
 
 const slots = useSlots();
+const attrs = useAttrs();
+
+const computedVariant = computed(() => {
+  if (props.variant) return props.variant;
+  if (attrs.solid !== undefined) return 'solid';
```
Comment (pranavrajs) [inline app/javascript/dashboard/components-next/button/Button.vue:42]: Is it required to check this with undefined? Shouldn't `attrs.solid` should work?

Comment (pranavrajs) [inline app/javascript/dashboard/components-next/button/Button.vue:42]:
```suggestion
  if (attrs.solid) return 'solid';
```

Is it mandatory to check this with `undefined`? Shouldn’t `attrs.solid` suffice?

**Comment by scmmishra** [comment]
@pranavrajs fixed the slot issue

File: app/javascript/dashboard/components-next/button/Button.vue (hunk 4 above)
Comment (scmmishra) [reply app/javascript/dashboard/components-next/button/Button.vue:42]: Fixed it here: https://github.com/chatwoot/chatwoot/pull/10473

**Review by iamsivin — COMMENTED** [review]
@scmmishra I tested the filter PR on staging and found a couple of minor issues. I’ve detailed them in the attached Loom videos. Could you please take a look?

https://www.loom.com/share/0f9924095080496db41a297e6f28aa43?sid=67507bf1-9622-4142-a4f7-b3b53e70ec27
https://www.loom.com/share/b7b331a02e804499ac5751f17527dd1d?sid=49d0f73e-ba2b-481b-b286-7560d7912d3e

File: app/javascript/dashboard/components-next/filter/inputs/MultiSelect.vue [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    <template #trigger="{ toggle }">
+      <button
+        v-if="hasItems"
+        class="bg-n-alpha-2 py-2 rounded-lg h-8 flex items-center px-0"
+        @click="toggle"
+      >
+        <div
+          v-for="item in selectedVisibleItems"
+          :key="item.name"
+          class="px-3 border-r border-n-weak text-n-slate-12 text-sm flex gap-2 items-center max-w-[100px]"
```
Comment (iamsivin) [inline app/javascript/dashboard/components-next/filter/inputs/MultiSelect.vue:100]: Let us consider the border for RTL also.
image

File: app/javascript/dashboard/components-next/filter/ConversationFilter.vue [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+  }
+};
+
+const addFilter = () => {
+  filters.value.push({
+    attribute_key: 'status',
+    filter_operator: 'equal_to',
+    values: '',
+    query_operator: 'and',
+  });
```
Comment (iamsivin) [inline app/javascript/dashboard/components-next/filter/ConversationFilter.vue:58]: Can we move it to a constant and reuse it here?

File: app/javascript/dashboard/components-next/filter/ConversationFilter.vue [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+const { t } = useI18n();
+const store = useStore();
+
+const resetFilter = () => {
+  filters.value = [
+    {
+      attribute_key: 'status',
+      filter_operator: 'equal_to',
+      values: '',
+      query_operator: 'and',
```
Comment (iamsivin) [inline app/javascript/dashboard/components-next/filter/ConversationFilter.vue:39]: Can we move it to a constant and use it here?

File: app/javascript/dashboard/components-next/filter/inputs/MultiSelect.vue [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+        >
+          <Icon v-if="item.icon" :icon="item.icon" class="flex-shrink-0" />
+          <span class="truncate">{{ item.name }}</span>
+        </div>
+        <div
+          v-if="remainingItems.length > 0"
+          v-tooltip.top="remainingTooltip"
+          class="px-3 border-r border-n-weak text-n-slate-12 text-sm flex gap-2 items-center max-w-[100px]"
```
Comment (iamsivin) [inline app/javascript/dashboard/components-next/filter/inputs/MultiSelect.vue:108]: Let us consider the border for RTL also.

File: app/javascript/dashboard/components-next/filter/ConversationFilter.vue [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+  emit('close');
+};
+
+const filterModalHeaderTitle = computed(() => {
+  return !props.isFolderView
+    ? t('FILTER.TITLE')
+    : t('FILTER.EDIT_CUSTOM_FILTER');
+});
+
+onMounted(() => dialogRef.value.open());
```
Comment (iamsivin) [inline app/javascript/dashboard/components-next/filter/ConversationFilter.vue:104]: Could you help me understand why we need to open the dialog on mounted?

**Comment by pranavrajs** [comment]
@scmmishra Here is my feedback: A couple of improvements could reduce the effort required to use the same in the other parts of the product. Please see my thoughts below.

__Make ConditionRow reusable__: The ConditionsRow component is currently tightly coupled with the conversation context, making it specific to that use case and unusable for other sections like the contacts filter or automations. We should make it generic. 

A good approach would be to pass filterTypes as a prop to the component, which removes the dependency on the conversation context. Since the component will no longer rely on the store, you can write the stories for the same for testing it on Histoire.

__Refactoring ConversationFilter component__: The conversation filter component can also benefit from being refactored into a container-component pattern. The container would manage interactions with the store and isolate the associated actions, while the component FilterForm could focus on rendering a reusable, generic form. This form should handle all form-related logic, including validations, and provide a callback to indicate when the form values are valid. This way, the FilterForm can be reused in other contexts, such as the contacts page, with minimal adjustments—only the container and the appropriate filterTypes would need to be updated.

File: app/javascript/dashboard/components-next/filter/provider.js [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+   */
+  const customFilterTypes = computed(() => {
+    return conversationAttributes.value.map(attr => {
+      return {
+        attribute_key: attr.attributeKey,
+        value: attr.attributeKey,
+        attribute_name: attr.attributeDisplayName,
+        label: attr.attributeDisplayName,
+        input_type: customAttributeInputType(attr.attributeDisplayType),
+        filter_operators: getOperatorTypes(attr.attributeDisplayType),
```
Comment (pranavrajs) [inline app/javascript/dashboard/components-next/filter/provider.js:99]: Is there a specific reason why we still use the snake case here? Couldn’t this be transformed before passing it to the component?

File: app/javascript/dashboard/components-next/filter/provider.js [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+      label: t('FILTER.ATTRIBUTES.LAST_ACTIVITY'),
+      input_type: 'date',
+      data_type: 'text',
+      filter_operators: dateOperators.value,
+      attribute_model: 'standard',
+    },
+    ...customFilterTypes.value,
+  ]);
+
+  const fitlerGroups = computed(() => {
```
Comment (pranavrajs) [inline app/javascript/dashboard/components-next/filter/provider.js:275]: Spelling mistake.

**Review by pranavrajs — CHANGES_REQUESTED** [review]
Please see the comments above.

File: app/javascript/dashboard/components-next/dialog/Dialog.vue [hunk 12]
```diff
@@ -44,6 +44,11 @@ const props = defineProps({
     type: Boolean,
     default: true,
   },
+  width: {
```
Comment (pranavrajs) [inline app/javascript/dashboard/components-next/dialog/Dialog.vue:47]: Seems like this is coming from a bad merge, please remove this.

**Comment by pranavrajs** [comment]
Some more feedback on the product side:

1/ I got an error where the value was displayed as “[Object object]”. However, I was not able to reproduce this issue again.
Screenshot 2024-11-24 at 8 26 48 PM
 
2/ I'd expect the dropdown to close automatically if I select a filterAttribute. Let us say I select Conversation Identifier from the list, there shouldn't be another click to close that list. See the interaction below.

https://www.loom.com/share/4ad94937b6ea488d854348e8726b0a9d

File: app/javascript/dashboard/components-next/filter/provider.js [hunk 13]
```diff
@@ -0,0 +1,299 @@
...
+  /**
+   * @type {import('vue').ComputedRef<FilterType[]>}
+   */
+  const filterTypes = computed(() => [
+    {
+      attribute_key: 'status',
+      value: 'status',
+      attribute_name: t('FILTER.ATTRIBUTES.STATUS'),
+      label: t('FILTER.ATTRIBUTES.STATUS'),
+      input_type: 'multiSelect',
```
Comment (pranavrajs) [inline app/javascript/dashboard/components-next/filter/provider.js:118]: status is implemented as single select in the backend. This is why there is an `all` key.

File: app/javascript/dashboard/components-next/filter/provider.js (hunk 10 above)
Comment (pranavrajs) [inline app/javascript/dashboard/components-next/filter/provider.js:99]: Is there a specific reason why we still use snake case attributes? Couldn’t we transform the keys while passing them to the component and then retrieve them from the component on the parent?

File: app/javascript/dashboard/components-next/filter/ConversationFilter.vue (hunk 9 above)
Comment (scmmishra) [reply app/javascript/dashboard/components-next/filter/ConversationFilter.vue:99]: The default dialog state is closed. So we manually have to open it. It's fixed in the popover PR. There we don't need this

**Comment by scmmishra** [comment]
> 1/ I got an error where the value was displayed as “[Object object]”. However, I was not able to reproduce this issue again.

Fixed this.



> I'd expect the dropdown to close automatically if I select a filterAttribute. Let us say I select Conversation Identifier from the list, there shouldn't be another click to close that list. See the interaction below.

Strange, this is not happening for me, probably an old commit? Can you check again?
